File: python/save_ldlfc_metrics.py
import numpy as np
from ldl_flc import LDL_FLC, fuzzy_cmeans, solve_LDM
from ldl_metrics import score
from util import load_dict
import csv
import sys


def run_and_save(dataset: str, out_csv: str):
    print(dataset)
    X, Y = np.load(dataset + "//feature.npy"), np.load(dataset + "//label.npy")
    train_inds = load_dict(dataset, "train_inds")
    test_inds = load_dict(dataset, "test_inds")

    rows = [("cheby","clark","can","kl","cosine","inter")]
    for i in range(10):
        print('training ' + str(i + 1) + ' fold')
        train_x, train_y = X[train_inds[i]], Y[train_inds[i]]
        test_x, test_y = X[test_inds[i]], Y[test_inds[i]]

        l1 = 0.001; l2 = 0.01; g = 5
        U = fuzzy_cmeans(train_y, g)
        manifolds = solve_LDM(train_y, U)

        ldl_flc = LDL_FLC(g, l1, l2)
        ldl_flc.fit(train_x, train_y, U, manifolds)
        ldl_flc.solve()

        cheby, clark, can, kl, cosine, inter = score(test_y, ldl_flc.predict(test_x))
        rows.append((cheby, clark, can, kl, cosine, inter))

    with open(out_csv, 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerows(rows)
    print('Saved Python metrics to', out_csv)


if __name__ == "__main__":
    ds = sys.argv[1] if len(sys.argv) > 1 else 'SJAFFE'
    out = sys.argv[2] if len(sys.argv) > 2 else ds + '/ldlfc_python_metrics.csv'
    run_and_save(ds, out)

File: matlab/save_ldlfc_metrics.m
function T = save_ldlfc_metrics(dataset, out_csv)
% Run LDL-FC on a dataset and save per-fold metrics to CSV
% Columns: cheby, clark, can, kl, cosine, inter
if nargin<1, dataset = 'python/SJAFFE'; end
if nargin<2, out_csv = fullfile(dataset, 'ldlfc_matlab_metrics.csv'); end
init_path(); % Initialize paths to core modules

X = util('load_npy', fullfile(dataset, 'feature.npy'));
Y = util('load_npy', fullfile(dataset, 'label.npy'));
train_inds = util('load_dict', dataset, 'train_inds');
test_inds = util('load_dict', dataset, 'test_inds');

M = zeros(10,6);
for i=1:10
    tr = train_inds{i}; te = test_inds{i};
    train_x = X(tr+1, :); train_y = Y(tr+1, :);
    test_x = X(te+1, :);  test_y = Y(te+1, :);
    mdl = LDL_FLC(5, 1e-3, 1e-2);
    mdl.fit(train_x, train_y);
    mdl.solve();
    Yhat = mdl.predict(test_x);
    [cheby, clark, can, kl, cosine, inter] = ldl_metrics('score', test_y, Yhat);
    M(i,:) = [cheby, clark, can, kl, cosine, inter];
end
T = array2table(M, 'VariableNames', {'cheby','clark','can','kl','cosine','inter'});
writetable(T, out_csv);
fprintf('Saved MATLAB metrics to %s\n', out_csv);
end
